tests/test_georaster.py

```diff
@@ -195,6 +196,162 @@ def test_loading(self) -> None:
+    )  # type: ignore
+    def test_data_setter(self, dtype: str, nodata_init: str | None) -> None:
+        """
+        Test that the behaviour of data setter, which is triggered directly using from_array, is as expected.
+        In details, we check that the data setter:
+
+        1. Writes the data in a masked array, whether the input is a classic array or a masked_array,
+        2. Reshapes the data in a 3D array if it is 2D,
+        3. Sets a new nodata value only if the provided array is not a masked_array and contains non-finite values,
+        4. Masks non-finite values that are unmasked, whether the input is a classic array or a masked_array,
+        """
+
+        nodata_init = None
+        dtype = "float32"
+
+        # Initiate a random array for testing
+        width = height = 5
+        transform = rio.transform.from_bounds(0, 0, 1, 1, width, height)
+
+        # Create random values between the lower and upper limit of the data type, max absolute 99999 for floats
+        if "int" in dtype:
+            val_min = np.iinfo(int_type=dtype).min
+            val_max = np.iinfo(int_type=dtype).max
+            randint_dtype = dtype
+        else:
+            val_min = -99999
+            val_max = 99999
+            randint_dtype = "int32"
+
+        # Fix the random seed
+        np.random.seed(42)
+        arr = np.random.randint(low=val_min, high=val_max, size=(1, width, height), dtype=randint_dtype).astype(dtype)
+        mask = np.random.randint(0, 2, size=(1, width, height), dtype=bool)
+
+        # Check that we are actually masking stuff
+        assert np.count_nonzero(mask) > 0
+
+        # Add a random floating point value if the data type is float
+        if "float" in dtype:
+            arr += np.random.normal(size=(1, width, height))
+
+        # Use either the default nodata or None
+        if nodata_init == "default":
+            nodata: int | None = _default_ndv(dtype)
+        else:
+            nodata = None
+
+        # -- First test: consistency with input array --
+
+        # 3 cases: classic array without mask, masked_array without mask and masked_array with mask
+        r1 = gr.Raster.from_array(data=arr, transform=transform, crs=None, nodata=nodata)
+        r2 = gr.Raster.from_array(data=np.ma.masked_array(arr), transform=transform, crs=None, nodata=nodata)
+        r3 = gr.Raster.from_array(data=np.ma.masked_array(arr, mask=mask), transform=transform, crs=None, nodata=nodata)
+
+        # Check nodata is correct
+        assert r1.nodata == nodata
+        assert r2.nodata == nodata
+        assert r3.nodata == nodata
+
+        # Compare the consistency of the data setter whether it is passed a masked_array or an unmasked one
+        assert np.array_equal(r1.data.data, arr)
+        assert not r1.data.mask
+        assert np.array_equal(r2.data.data, arr)
+        assert not r2.data.mask
+        assert np.array_equal(r3.data.data, arr)
```

Comment:
`gu.misc.array_equal` checks that the array **and** mask are equal.

Reply:
Actually it uses the mask to replace masked values with NaNs, so this functionality already exists as `np.ma.allequal`. I'm not sure there any added value to it now that we are always using NumPy > 1.20 that has good internal functions. Adding this to the related issue for closing.

Regarding the current case: it cannot be handled by a single NumPy (or GeoUtils) function:  the mask is not defined, so `data.mask` is `False` without any shape. I think it can only be compared this way, which is not too much trouble.